src/nvim/popupmnu.c

```diff
@@ -537,24 +537,30 @@ void pum_redraw(void)
                 }
               }
               grid_puts_len(&pum_grid, rt, (int)STRLEN(rt), row,
-                            col - size + 1, attr);
+                            grid_col - size + 1, attr);
               xfree(rt_start);
               xfree(st);
-              col -= width;
+              grid_col -= width;
             } else {
               int size = (int)STRLEN(st);
               int cells = (int)mb_string2cells(st);
 
               // only draw the text that fits
-              while (size > 0 && col + cells > pum_width + pum_col) {
+              while (size > 0 && grid_col - col_off + cells > pum_width) {
                 size--;
                 size -= utf_head_off(st, st + size);
                 cells -= utf_ptr2cells(st + size);
               }
 
-              grid_puts_len(&pum_grid, st, size, row, col, attr);
+              grid_puts_len(&pum_grid, st, size, row, grid_col, attr);
               xfree(st);
-              col += width;
+              if (grid_col - col_off + cells < pum_width) {
+                // Most right character requires 2-cells but only 1 cell
+                // is available on screen.  Put a '<' on the right of the
+                // pum item
+                grid_puts_len(&pum_grid, (char_u *)">", 1, row, grid_col + cells, attr);
+              }
```

Comment:
Making this deviation from Vim makes the test I wrote pass. Otherwise popup menu fails to show in test. Maybe it is because of `NUL` characters.